### PR TITLE
Add dashvector self query retriever

### PR DESCRIPTION
**Pull request by xiaoyuxee**
## Description
Add `Dashvector` retriever and self-query retriever

## How to use
```python
from langchain.vectorstores.dashvector import DashVector

vectorstore = DashVector.from_documents(docs, embeddings)
retriever = SelfQueryRetriever.from_llm(
    llm, vectorstore, document_content_description, metadata_field_info, verbose=True
)
```

### CODE REVIEW

File: libs/langchain/langchain/retrievers/dashvector.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+from langchain.callbacks.manager import (
+    AsyncCallbackManagerForRetrieverRun,
+    CallbackManagerForRetrieverRun,
+)
+from langchain.docstore.document import Document
+from langchain.embeddings.base import Embeddings
+from langchain.schema import BaseRetriever
+
+
+class DashVectorRetriever(BaseRetriever):
```
Comment (baskaryan) [inline libs/langchain/langchain/retrievers/dashvector.py:19]: is there a different between this class and the functionality you'd get with DashVector.as_retriever()?

Comment (xiaoyuxee) [reply libs/langchain/langchain/retrievers/dashvector.py]: No. Thanks for your reminder, I have removed it and add a `DashVectorHybridSearchRetriever` when this  feature is supported in the future.

**Comment by xiaoyuxee** [comment]
@baskaryan can you review the response to the comments and let me know whether this pr is ready for merge? thanks

**Review by hwchase17 — APPROVED** [review]
looks reasonable - thanks!

**Pull request merged**